### PR TITLE
Support uint256 upkeep ids everywhere in codebase

### PR DESCRIPTION
**Pull request by RyanRHall**

### CODE REVIEW

File: core/chains/evm/txmgr/models.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ type EthTxMeta struct {
 	SubID *uint64 `json:"SubId,omitempty"`
 
 	// Used for keepers
-	UpkeepID *int64 `json:"UpkeepID,omitempty"`
+	UpkeepID *string `json:"UpkeepID,omitempty"`
```
Comment (infiloop2) [inline core/chains/evm/txmgr/models.go:44]: why not bigInt here

File: core/services/keeper/orm.go [hunk 2]
```diff
@@ -106,36 +107,47 @@ DELETE FROM upkeep_registrations WHERE registry_id IN (
 	return rowsAffected, nil
 }
 
-//EligibleUpkeepsForRegistry fetches eligible upkeeps for processing
-//The query checks the following conditions
-// - checks the registry address is correct and the registry has some keepers associated
-// -- is it my turn AND my keeper was not the last perform for this upkeep OR my keeper was the last before BUT it is past the grace period
-// -- OR is it my buddy's turn AND they were the last keeper to do the perform for this upkeep
+/*
```
Comment (infiloop2) [inline core/services/keeper/orm.go:110]: wasn't able to comment on unchanged lines, there are more functions that need changing: BatchDeleteUpkeepsForJob

File: core/services/keeper/registry_synchronizer_sync.go [hunk 3]
```diff
@@ -167,9 +166,9 @@ func (rs *RegistrySynchronizer) newRegistryFromChain() (Registry, error) {
 
 // CalcPositioningConstant calculates a positioning constant.
 // The positioning constant is fixed because upkeepID and registryAddress are immutable
-func CalcPositioningConstant(upkeepID int64, registryAddress ethkey.EIP55Address) (int32, error) {
+func CalcPositioningConstant(upkeepID *utils.Big, registryAddress ethkey.EIP55Address) (int32, error) {
 	upkeepBytes := make([]byte, binary.MaxVarintLen64)
-	binary.PutVarint(upkeepBytes, upkeepID)
+	binary.PutVarint(upkeepBytes, upkeepID.ToInt().Int64()) // XXX - not valid but doesn't matter bc we're removing this code
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_sync.go:171]: create a ticket for this?

File: core/services/keeper/upkeep_executer.go [hunk 4]
```diff
@@ -219,6 +218,7 @@ func (ex *UpkeepExecuter) execute(upkeep UpkeepRegistration, head *evmtypes.Head
 			"fromAddress":           upkeep.Registry.FromAddress.String(),
 			"contractAddress":       upkeep.Registry.ContractAddress.String(),
 			"upkeepID":              upkeep.UpkeepID,
```
Comment (infiloop2) [inline core/services/keeper/upkeep_executer.go:220]: do we need this if we have prettyID?

File: core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql [hunk 5]
```diff
@@ -0,0 +1,48 @@
+-- +goose Up
```
Comment (infiloop2) [inline core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql:1]: will also need to update the job spec for pretyID

File: core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  max_bigint NUMERIC := '9223372036854775808' :: NUMERIC(19, 0);
+  result BIT VARYING;
+BEGIN
+  ASSERT num <= 115792089237316195423570985008687907853269984665640564039457584007913129639935, "value larger than max uint256";
+  WITH
+      chunks (exponent, chunk) AS (
+        SELECT
+          exponent,
+          floor((num / (max_bigint ^ exponent) :: NUMERIC(256, 20)) % max_bigint) :: BIGINT
+        FROM generate_series(0, 5) exponent
```
Comment (infiloop2) [inline core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql:20]: This went over my head. Could you add some comments explaining what's happening here

File: core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    )
+  SELECT bit_or(chunk :: BIT(256) :: BIT VARYING << (63 * (exponent))) :: BIT VARYING
+  FROM chunks INTO result;
+  RETURN result;
```
Comment (infiloop2) [inline core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql:24]: This went over my head. Could you add some comments explaining what's happening here

File: core/chains/evm/txmgr/models.go (hunk 1 above)
Comment (RyanRHall) [reply core/chains/evm/txmgr/models.go:44]: this is the hex-encoded upkeep ID, ex `UPx1234...ABCD`

File: core/services/keeper/upkeep_executer.go (hunk 4 above)
Comment (RyanRHall) [reply core/services/keeper/upkeep_executer.go:220]: we do because the job pipeline expects a bigInt for encoding checkUpkeep & performUpkeep payloads

File: core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql (hunk 5 above)
Comment (RyanRHall) [reply core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql:1]: prettyID is derived directly from upkeepID, so it doesn't need a column in the DB

File: core/services/keeper/registry_synchronizer_sync.go (hunk 3 above)
Comment (RyanRHall) [reply core/services/keeper/registry_synchronizer_sync.go:171]: found a simple fix

File: core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+CREATE OR REPLACE FUNCTION least_significant(bits BIT VARYING, num integer)
+  RETURNS BIT VARYING AS $$
+BEGIN
+  ASSERT length(bits) >= num, "slice is larger than input";
+  RETURN substring(bits from length(bits) - num + 1 for num);
+END;
+$$ LANGUAGE plpgsql;
+
+
+-- CREATE UNIQUE INDEX idx_upkeep_registrations_unique_upkeep_ids_per_keeper ON upkeep_registrations(registry_id, upkeep_id);
```
[the remark this answers is not in the document]
Comment (RyanRHall) [reply core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql:38]: nope!

File: core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql (hunk 5 above)
Comment (infiloop2) [reply core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql:1]: I meant updating the observation source in job spec.

to change ```txMeta="{\\"jobID\\":$(jobSpec.jobID),\\"upkeepID\\":$(jobSpec.upkeepID)}"```
to ```txMeta="{\\"jobID\\":$(jobSpec.jobID),\\"upkeepID\\":$(jobSpec.prettyID)}"``` within DB

File: core/services/keeper/upkeep_executer.go (hunk 4 above)
Comment (RyanRHall) [reply core/services/keeper/upkeep_executer.go:220]: the pipeline isn't that flexible, we're basically telling it "use upkeepID for sending txs" and "use prettyID for logging" -  imho, better that the pipeline is agnostic about keeper intricacies like that

File: core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql (hunk 5 above)
Comment (RyanRHall) [reply core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql]: yes! good catch.

File: core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql (hunk 7 above)
Comment (RyanRHall) [reply core/store/migrate/migrations/0115_migrate_upkeep_id_to_big_int.sql]: added comments

File: core/services/keeper/orm.go (hunk 2 above)
Comment (RyanRHall) [reply core/services/keeper/orm.go]: updated!

File: core/services/keeper/orm.go [hunk 9]
```diff
@@ -112,31 +113,42 @@ DELETE FROM upkeep_registrations WHERE registry_id IN (
...
-                      CAST($4 AS bit(32)))::bigint % keeper_registries.num_keepers)
-            AND
-                    upkeep_registrations.last_keeper_index IS NOT DISTINCT FROM (keeper_registries.keeper_index + 1) % keeper_registries.num_keepers
-        ))
+SELECT upkeep_registrations.*
+FROM upkeep_registrations
+  INNER JOIN keeper_registries ON keeper_registries.id = upkeep_registrations.registry_id,
+  LATERAL ABS(
+		(least_significant(uint256_to_bit(upkeep_registrations.upkeep_id), 32) # least_significant($4, 32))::bigint
+	) AS turn
```
Comment (infiloop2) [inline core/services/keeper/orm.go:125]: Nice!

File: core/services/keeper/orm.go [hunk 10]
```diff
@@ -91,7 +92,7 @@ RETURNING *
 }
 
 // BatchDeleteUpkeepsForJob deletes all upkeeps by the given IDs for the job with the given ID
-func (korm ORM) BatchDeleteUpkeepsForJob(jobID int32, upkeepIDs []int64) (int64, error) {
+func (korm ORM) BatchDeleteUpkeepsForJob(jobID int32, upkeepIDs []utils.Big) (int64, error) {
```
Comment (infiloop2) [inline core/services/keeper/orm.go:95]: Not sure if it makes a difference, why not `upkeepIDs []*utils.Big` here

File: core/services/keeper/orm_test.go [hunk 11]
```diff
@@ -359,3 +361,99 @@ func TestKeeperDB_NewSetLastRunInfoForUpkeepOnJob(t *testing.T) {
...
+			name:  "max",
+			input: max,
+		},
+		{
+			name:  "rand",
+			input: rand,
+		},
+		{
+			name:  "padded",
+			input: min,
```
Comment (infiloop2) [inline core/services/keeper/orm_test.go:438]: does this test anything different from 'min' test case?

File: core/store/migrate/migrations/0116_migrate_upkeep_id_to_big_int.sql [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    SELECT
+      exponent,
+      floor(num::NUMERIC(178,100) / (max_int32 ^ exponent) % max_int32)::BIGINT from generate_series(0,7) exponent
+  )
+  -- concat 32 bit chunks together
+  SELECT bit_or(chunk::bit(256) << (32*(exponent)))
+  FROM chunks INTO result;
+  RETURN result;
+END;
+$$ LANGUAGE plpgsql;
```
Comment (infiloop2) [inline core/store/migrate/migrations/0116_migrate_upkeep_id_to_big_int.sql:64]: looks good! if only this was a native function within psql

File: core/utils/big.go [hunk 13]
```diff
@@ -149,6 +150,11 @@ func (b *Big) Hex() string {
 	return hexutil.EncodeBig(b.ToInt())
 }
 
+// Hex returns the hex encoding of b.
```
Comment (infiloop2) [inline core/utils/big.go:153]: outdated comment

File: core/services/keeper/upkeep_executer_test.go [hunk 14]
```diff
@@ -119,12 +119,13 @@ func Test_UpkeepExecuter_PerformsUpkeep_Happy(t *testing.T) {
 	t.Run("runs upkeep on triggering block number", func(t *testing.T) {
 		db, config, ethMock, executer, registry, upkeep, job, jpv2, txm, _, _, _ := setup(t)
 
-		gasLimit := upkeep.ExecuteGas + config.KeeperRegistryPerformGasOverhead()
+		// gasLimit := upkeep.ExecuteGas + config.KeeperRegistryPerformGasOverhead()
```
Comment (infiloop2) [inline core/services/keeper/upkeep_executer_test.go:122]: changes in this file look unrelated

**Review by infiloop2 — CHANGES_REQUESTED** [review]
Looking good, added some minor comments. Also tests need fixing

File: core/services/keeper/orm_test.go (hunk 11 above)
Comment (RyanRHall) [reply core/services/keeper/orm_test.go:438]: oops, my B!

Comment (RyanRHall) [reply core/services/keeper/orm_test.go:438]: okay, fixed

File: core/store/migrate/migrations/0116_migrate_upkeep_id_to_big_int.sql (hunk 12 above)
Comment (RyanRHall) [reply core/store/migrate/migrations/0116_migrate_upkeep_id_to_big_int.sql:64]: I know, this was quite the doozey for me. learned a  lot about postgres numeric types in the process. Should also point out if we need to optimize the turn taking query, this could probably be replaced by us just storing the least significant 32 bits of the upkeepIDs

File: core/services/keeper/orm.go (hunk 10 above)
Comment (infiloop2) [reply core/services/keeper/orm.go:95]: interesting, TIL!

**Review by infiloop2 — CHANGES_REQUESTED** [review]
looks good. Smoke tests seem to be failing with this error:

```
Creating KeeperV2 Job shouldn't fail
[1393](https://github.com/smartcontractkit/chainlink/runs/6216991740?check_suite_focus=true#step:10:1393)
  Unexpected error:
[1394](https://github.com/smartcontractkit/chainlink/runs/6216991740?check_suite_focus=true#step:10:1394)
      <*errors.errorString | 0xc00c6b5300>: {
[1395](https://github.com/smartcontractkit/chainlink/runs/6216991740?check_suite_focus=true#step:10:1395)
          s: "unexpected response code, got 400, expected 200\nURL: http://localhost:43091\nresponse received: {\"errors\":[{\"detail\":\"invalid observation source provided\"}]}",
[1396](https://github.com/smartcontractkit/chainlink/runs/6216991740?check_suite_focus=true#step:10:1396)
      }
[1397](https://github.com/smartcontractkit/chainlink/runs/6216991740?check_suite_focus=true#step:10:1397)
      unexpected response code, got 400, expected 200
[1398](https://github.com/smartcontractkit/chainlink/runs/6216991740?check_suite_focus=true#step:10:1398)
      URL: http://localhost:43091
[1399](https://github.com/smartcontractkit/chainlink/runs/6216991740?check_suite_focus=true#step:10:1399)
      response received: {"errors":[{"detail":"invalid observation source provided"}]}
[1400](https://github.com/smartcontractkit/chainlink/runs/6216991740?check_suite_focus=true#step:10:1400)
  occurred
```

**Comment by infiloop2** [comment]
Ok we'll need to change https://github.com/smartcontractkit/integrations-framework/blob/main/client/chainlink_models.go#L1038 for the smoke tests to pass. 

@kalverra what's the recommended process to resolve this cross repo dependency? Merge this PR first or merge PR in integrations-framework first? 

Also is there any way we can run smoke tests on this PR against an updated version of integrations-framework repo to ensure nothing else breaks before merging?
